x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java

```diff
@@ -0,0 +1,513 @@
+            final IndexNameExpressionResolver indexNameExpressionResolver,
+            final RemoteClusterService remoteClusterService,
+            final RemoteClusterLicenseChecker remoteClusterLicenseChecker,
+            final String[] source,
+            final String dest,
+            final String nodeName,
+            final String license
+        ) {
+            this.state = state;
+            this.indexNameExpressionResolver = indexNameExpressionResolver;
+            this.remoteClusterService = remoteClusterService;
+            this.remoteClusterLicenseChecker = remoteClusterLicenseChecker;
+            this.source = source;
+            this.dest = dest;
+            this.nodeName = nodeName;
+            this.license = license;
+        }
+
+        public ClusterState getState() {
+            return state;
+        }
+
+        public RemoteClusterService getRemoteClusterService() {
+            return remoteClusterService;
+        }
+
+        public RemoteClusterLicenseChecker getRemoteClusterLicenseChecker() {
+            return remoteClusterLicenseChecker;
+        }
+
+        public IndexNameExpressionResolver getIndexNameExpressionResolver() {
+            return indexNameExpressionResolver;
+        }
+
+        public boolean isRemoteSearchEnabled() {
+            return remoteClusterLicenseChecker != null;
+        }
+
+        public String[] getSource() {
+            return source;
+        }
+
+        public String getDest() {
+            return dest;
+        }
+
+        public String getNodeName() {
+            return nodeName;
+        }
+
+        public String getLicense() {
+            return license;
+        }
+
+        public Set<String> resolveSource() {
+            if (resolvedSource == null) {
+                resolveLocalAndRemoteSource();
+            }
+
+            return resolvedSource;
+        }
+
+        public Set<String> resolveRemoteSource() {
+            if (resolvedRemoteSource == null) {
+                resolveLocalAndRemoteSource();
+            }
+
+            return resolvedRemoteSource;
+        }
+
+        public String resolveDest() {
+            if (resolvedDest == null) {
+                try {
+                    Index singleWriteIndex = indexNameExpressionResolver.concreteWriteIndex(
+                        state,
+                        IndicesOptions.lenientExpandOpen(),
+                        dest,
+                        true
+                    );
+                    if (singleWriteIndex != null) {
+                        resolvedDest = singleWriteIndex.getName();
+                    } else {
+                        resolvedDest = dest;
+                    }
+                } catch (IllegalArgumentException e) {
+                    // stop here as we can not return a single dest index
+                    addValidationError(e.getMessage(), false);
+                }
+            }
+
+            return resolvedDest;
```

Comment:
I think this outside return could be removed instead of mutating in the `try..` block

Reply:
`resolveDest` can be called repeatedly (by different validators) and resolves the source on the 1st call on-demand.